internal/storage/v1/elasticsearch/spanstore/reader.go

```
@@ -457,41 +444,67 @@ func (s *SpanReader) multiRead(ctx context.Context, traceIDs []model.TraceID, st
 				traceSpan.Spans = append(traceSpan.Spans, spans...)
 			} else {
-				tracesMap[lastSpan.TraceID] = &model.Trace{Spans: spans}
+				tracesMap[lastSpan.TraceID] = &dbmodel.Trace{Spans: spans}
 			}
 
 			totalDocumentsFetched[lastSpan.TraceID] += len(result.Hits.Hits)
 			if totalDocumentsFetched[lastSpan.TraceID] < int(result.TotalHits()) {
 				traceIDs = append(traceIDs, lastSpan.TraceID)
-				searchAfterTime[lastSpan.TraceID] = model.TimeAsEpochMicroseconds(lastSpan.StartTime)
+				searchAfterTime[lastSpan.TraceID] = lastSpan.StartTime
 			}
 		}
 	}
 
-	var traces []*model.Trace
+	var traces []*dbmodel.Trace
 	for _, t := range tracesMap {
 		traces = append(traces, t)
 	}
 	return traces, nil
 }
 
-func buildTraceByIDQuery(traceID model.TraceID) elastic.Query {
-	traceIDStr := traceID.String()
-
+func buildTraceByIDQuery(traceID dbmodel.TraceID) (elastic.Query, error) {
+	traceIDStr := string(traceID)
 	if traceIDStr[0] != '0' || disableLegacyIDs.IsEnabled() {
-		return elastic.NewTermQuery(traceIDField, traceIDStr)
+		return elastic.NewTermQuery(traceIDField, traceIDStr), nil
 	}
 	// https://github.com/jaegertracing/jaeger/pull/1956 added leading zeros to IDs
 	// So we need to also read IDs without leading zeros for compatibility with previously saved data.
 	var legacyTraceID string
-	if traceID.High == 0 {
-		legacyTraceID = strconv.FormatUint(traceID.Low, 16)
+	hi, lo, err := getHiAndLoFromString(traceIDStr)
+	if err != nil {
+		return nil, err
+	}
+	if *hi == 0 {
+		legacyTraceID = strconv.FormatUint(*lo, 16)
 	} else {
-		legacyTraceID = fmt.Sprintf("%x%016x", traceID.High, traceID.Low)
+		legacyTraceID = fmt.Sprintf("%x%016x", *hi, *lo)
 	}
-	return elastic.NewBoolQuery().Should(
+	query := elastic.NewBoolQuery().Should(
 		elastic.NewTermQuery(traceIDField, traceIDStr).Boost(2),
 		elastic.NewTermQuery(traceIDField, legacyTraceID))
+	return query, nil
+}
+
+func getHiAndLoFromString(traceIDStr string) (high *uint64, low *uint64, er error) {
```

I couldn't find any other way for doing this (this code is exactly same as `TraceIDFromString`). Also this is going to be depreceated in coming releases. If this looks good, I will complete code coverage for it!

---

If dependency on `model/v1` is permissible then we can remove this function and use `NewTraceIDFromString`, use that trace id to extract legacy id!

---

Rather blindly copying the code, what do you actually want to achieve? And is the hi/lo distinction necessary for that goal?

Also, where is the feature gate used here?

---

As per me, yes it is necessary because we want to apply leading zeroes, so we need to convert them to integers. Regarding the feature gate, it was applied in the PR: https://github.com/jaegertracing/jaeger/pull/6848, it is a part of extracting legacy ids!

---

Why is it necessary to convert to integers to pad with zeroes?

---

Because we need to handle some edge cases, like when trace id is very long (length > 32) or when it is short (length < 16). This is what is handled by this method. Also in any case in between this the code becomes simple when we convert them to integers and then pad zeroes. If we will only use strings then I am concerned that we might miss some edge case. Therefore for simplicity I think we should convert this string to uint and I reached to this same algorithm when moved with this approach. So should we go with this or just use strings, without converting them to uint?

---

original code:
```
	if *hi == 0 {
		legacyTraceID = strconv.FormatUint(traceID.Low, 16)
	} else {
		legacyTraceID = fmt.Sprintf("%x%016x", traceID.High, traceID.Low)
	}
```
The ONLY thing this algorithm achieves is rendering an ID to a hex string without leading zeroes. If your input is already a string it means you simply need to remove leading zeros from it.

The len>32 case isn ot possible in previous code, but doesn't require special handling, not do we need to raise an error, we can just use the string as is.